### ISSUE TITLE
Race conditions when loading file / PDB id

### ISSUE THREAD

**Issue by vhorvath** [nglviewer/nglview#881]
Hello,

I am amazed by the amount of work that had been gone into the `NGLView` and the `nglviewer `widget, and I have been using it quite a bit lately. Very nice work!

I came across some inconsistencies concerning the communication between the JS and the Python sides of this project. I see that some JS calls are running in async with the Python side, but this yields some race conditions which are hard to tease out and yields inconsistent behaviors that are extremely hard to understand if you work on the Python side. (Once I noticed that I need to look at the JS Debug console things got a little better.)

If you take the simple Jupyter notebook example:
```
# Cell 1
import nglview as nv

# Cell 2
widget= nv.NGLWidget()
widget.add_pdbid('6urh')
widget.add_representation('licorice', component=0)
display(widget)
```

The licorice will not render. JS Debug console will show:
```
STAGE LOG loading file '6urh.pdb'
Exception in Comm callback TypeError: Cannot read property 'addRepresentation' of undefined`. 
~~~~
STAGE LOG loaded '6urh.pdb'
```

Which is due to adding a new representation to a non existent component. Whether `display()` is called first does not matter. Since the error message is flanked by the logs of the loading process `STAGE LOG` the representation must be added before loading finished.

To test if this is due to a race condition issue, I produced another example. Here we first download the file. If `Cell #2` runs a couple of times, you get intermittent errors.

```
# Cell 1
import nglview as nv
!wget http://files.rcsb.org/download/6urh.pdb

# Cell 2
widget = nv.NGLWidget()
widget.add_component('6urh.pdb')
widget.add_representation('licorice', component=0)
display(widget)
```

Sometimes you'll see the licorice sometimes you won't. When licorice is missing, then the same error as above appears in the JS Debug console.

I checked the source code and I think the intent is that `loadFile` remote calls are blocking, but for some reason the blocking loop is being passed by.

I don't have your tool chain set up at the moment for patching the JS code, but I managed to created a proof of concept patch for the problem. [index.js.diff.txt](https://github.com/arose/nglview/files/4357296/index.js.diff.txt) It is ugly, but I don't know my way around your JS code very well, I only wanted to see if my theory is right. By using the patch the race condition can be eliminated.

Basically, the `_handleLoadFileFinished()` method (which notifies the Python side) is called before the file is actually loaded. This is because the `stage.loadFile()` method results in a `Promise` which immediately returns and causes the notification to be sent to Python. I moved the `_handleLoadFileFinished()` call into the `onLoadFn()` which is called after the file load promise chain completes. This required some ugly object passing, I am sure there is a better way to reach the NGLView instance, but I could not see it.

**Comment by hainm** [comment]
Thanks. I will have a look.

**Comment by vhorvath** [comment]
Thanks Hai.

I have spent more time figuring out the parts of the `NGLView` project. I understand it a little better.

I think I found a more elegant solution to the problem above. It is as simple as adding an `await` in the `line 1054` in `widget_ngl.ts` for the method call yielding a `Promise`:

```var o = await this._getLoadFilePromise(msg)```

I tried this yesterday, but did not work (must have been Jupyter serving up the wrong `index.js`).

This solution is more in line with your previous methods, and it does not require the modification of the `ngl` package.

I will prepare a PR with it.
